src/Elf.cpp

```diff
@@ -0,0 +1,989 @@
+        uint64_t size = output.size() - offset;
+        append_padding(output, alignment);
+
+        Shdr<T> shdr;
+        safe_assign(shdr.sh_name, strings.get(".rela" + s.get_name()));
+        shdr.sh_type = Section::SHT_RELA;
+        shdr.sh_flags = Section::SHF_ALLOC;
+        safe_assign(shdr.sh_offset, offset);
+        safe_assign(shdr.sh_addr, offset);
+        safe_assign(shdr.sh_size, size);
+        safe_assign(shdr.sh_addralign, alignment);
+
+        safe_assign(shdr.sh_link, symtab_idx);
+        safe_assign(shdr.sh_info, section_idxs[&s]);
+        shdr.sh_entsize = sizeof(Rela<T>);
+
+        uint16_t shndx = safe_cast<uint16_t>(shdrs.size());
+        shdrs.push_back(shdr);
+        return shndx;
+    };
+
+    addr_t rela_got_idx = write_relocation_section(got);
+
+    // Add some strings we know we'll need in the string table after we write it.
+    strings.get(soname);
+    for (const auto &i : dependencies) {
+        strings.get(i);
+    }
+
+    Section dynamic(".dynamic", Section::SHT_DYNAMIC);
+    strings.get(dynamic.get_name());
+    dynamic.set_alignment(4);
+    dynamic.set_flag(Section::SHF_ALLOC);
+    Section strtab(".strtab", Section::SHT_STRTAB);
+    strings.get(strtab.get_name());
+    strtab.set_flag(Section::SHF_ALLOC);
+    strtab.set_contents(strings.table);
+    uint16_t strtab_idx = write_section(strtab);
+
+    std::vector<Dyn<T>> dyn;
+    auto make_dyn = [](int32_t tag, addr_t val = 0) {
+        Dyn<T> d;
+        d.d_tag = tag;
+        d.d_val = val;
+        return d;
+    };
+
+    for (const auto &i : dependencies) {
+        dyn.push_back(make_dyn(DT_NEEDED, strings.get(i)));
+    }
+    if (!soname.empty()) {
+        dyn.push_back(make_dyn(DT_SONAME, strings.get(soname)));
+    }
+    dyn.push_back(make_dyn(DT_SYMBOLIC));
+
+    // This is really required...
+    dyn.push_back(make_dyn(DT_HASH, get_section_offset(hash)));
+
+    // Address of the symbol table.
+    dyn.push_back(make_dyn(DT_SYMTAB, shdrs[dynsym_idx].sh_offset));
+    dyn.push_back(make_dyn(DT_SYMENT, shdrs[dynsym_idx].sh_entsize));
+
+    // Address of the string table.
+    dyn.push_back(make_dyn(DT_STRTAB, get_section_offset(strtab)));
+    dyn.push_back(make_dyn(DT_STRSZ, strtab.get_size()));
+
+    // Offset to the GOT.
+    dyn.push_back(make_dyn(DT_PLTGOT, get_section_offset(got)));
+
+    // Relocations associated with the PLT.
+    addr_t pltrelsz = sizeof(Rela<T>)*plt_symbols.size();
+    dyn.push_back(make_dyn(DT_JMPREL, shdrs[rela_got_idx].sh_offset));
+    dyn.push_back(make_dyn(DT_PLTREL, DT_RELA));
+    dyn.push_back(make_dyn(DT_PLTRELSZ, pltrelsz));
+
+    // Other relocations.
+    dyn.push_back(make_dyn(DT_RELA, shdrs[rela_got_idx].sh_offset + pltrelsz));
+    dyn.push_back(make_dyn(DT_RELASZ, shdrs[rela_got_idx].sh_size - pltrelsz));
+    dyn.push_back(make_dyn(DT_RELAENT, sizeof(Rela<T>)));
+
+    dynamic.set_contents(dyn);
+
+    // Add any target specific stuff.
+    linker->append_dynamic(dynamic);
```

Comment:
Is this order wrong? I noticed that the hexagon linker put the dynamic section early, but that requires writing it before some of the later addresses are known. It didn't seem like it was breaking anything to just put it at the end?